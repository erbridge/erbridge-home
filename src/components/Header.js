import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, { Component, Fragment } from 'react';
import { NavLink } from 'react-router-dom';

import Logo from './Logo';

import './Header.css';

class Header extends Component {
  static propTypes = {
    homeRoute: PropTypes.shape({
      exact: PropTypes.bool,
      path: PropTypes.string.isRequired,
    }).isRequired,
    pointerPosition: PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired,
    }),
    routes: PropTypes.arrayOf(
      PropTypes.shape({
        exact: PropTypes.bool,
        link: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        path: PropTypes.string.isRequired,
      }),
    ),
  };

  state = {
    navMenuIsOpen: false,
    shouldFocusLogo: false,
  };

  render() {
    const { homeRoute, pointerPosition, routes } = this.props;
    const { navMenuIsOpen, shouldFocusLogo } = this.state;

    return (
      <Fragment>
        <div
          className="Header"
          onFocus={() => this.setState({ shouldFocusLogo: true })}
          onBlur={() => this.setState({ shouldFocusLogo: false })}
          onMouseEnter={() => this.setState({ shouldFocusLogo: true })}
          onMouseLeave={() => this.setState({ shouldFocusLogo: false })}
        >
          <div className="Header__content">
            <NavLink exact={homeRoute.exact} to={homeRoute.path}>
              <Logo
                className="Header__logo"
                focused={shouldFocusLogo}
                pointerPosition={pointerPosition}
              />
            </NavLink>
            <div className="Header__spacer" />
            <div className="Header__nav">
              {routes &&
                routes
                  .map((route, index) => (
                    <NavLink
                      key={route.name || index}
                      className="Header__nav-link"
                      activeClassName="Header__nav-link--active"
                      exact={route.navExact}
                      to={route.path}
                    >
                      {route.link}
                    </NavLink>
                  ))
                  .reduce((elementsSoFar, link, index, links) => {
                    const elements = [...elementsSoFar, link];

                    if (index < links.length - 1) {
                      elements.push(
                        <div key={index} className="Header__nav-spacer" />,
                      );
                    }

                    return elements;
                  }, [])}
            </div>
            <div
              className={classnames('Header__nav-menu-trigger', {
                'Header__nav-menu-trigger--active': navMenuIsOpen,
              })}
              onClick={() => this.setState({ navMenuIsOpen: !navMenuIsOpen })}
            >
              Menu
              <div className="Header__nav-menu-icon" />
            </div>
          </div>
        </div>
        <div
          className={classnames('Header__nav-menu', {
            'Header__nav-menu--active': navMenuIsOpen,
          })}
        >
          <div className="Header__nav-menu__content">
            {routes &&
              routes.map((route, index) => (
                <NavLink
                  key={route.name || index}
                  className="Header__nav-menu-link"
                  activeClassName="Header__nav-menu-link--active"
                  exact={route.navExact}
                  to={route.path}
                  onClick={() => this.setState({ navMenuIsOpen: false })}
                  tabIndex={navMenuIsOpen ? 0 : -1}
                >
                  {route.link}
                </NavLink>
              ))}
          </div>
        </div>
      </Fragment>
    );
  }
}

export default Header;
