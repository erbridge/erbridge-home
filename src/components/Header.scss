@import 'styles/colours';
@import 'styles/layout';

@include wrapper-with-content(Header);

.Header {
  background-color: $grey;
  padding: 10px;
}

.Header__content {
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
}

.Header__logo {
  display: inherit;
  width: 110px;
}

.Header__spacer {
  flex: 1;
  min-width: 50px;
}

.Header__nav {
  display: flex;
  padding-bottom: 4px;
  flex-wrap: wrap;
  justify-content: center;
  text-align: right;
}

.Header__nav-spacer {
  flex: 1;
  min-width: 20px;
  max-width: 30px;
}

.Header__nav-link {
  margin-top: 10px;
  font-family: Dosis, sans-serif;
  font-weight: 400;
  font-size: 26px;
  line-height: 1;

  &.Header__nav-link--active:not(:active):not(:focus):not(:hover) {
    color: $white;
  }
}
