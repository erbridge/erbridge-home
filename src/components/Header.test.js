import { shallow } from 'enzyme';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import Header from './Header';

it('renders correctly', () => {
  expect(
    shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: '/' }]}>
        <Header homeRoute={{ path: '/' }} />
      </MemoryRouter>,
    ),
  ).toMatchSnapshot();
});
