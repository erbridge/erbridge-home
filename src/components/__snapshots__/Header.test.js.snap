// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
<Fragment>
  <div
    className="Header"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="Header__content"
    >
      <NavLink
        exact={true}
        to="/"
      >
        <Logo
          className="Header__logo"
          focused={false}
        />
      </NavLink>
      <div
        className="Header__spacer"
      />
      <div
        className="Header__nav"
      >
        <NavLink
          activeClassName="Header__nav-link--active"
          className="Header__nav-link"
          key="a"
          to="/a"
        >
          A
        </NavLink>
        <div
          className="Header__nav-spacer"
          key="0"
        />
        <NavLink
          activeClassName="Header__nav-link--active"
          className="Header__nav-link"
          key="b"
          to="/b"
        >
          B
        </NavLink>
        <div
          className="Header__nav-spacer"
          key="1"
        />
        <NavLink
          activeClassName="Header__nav-link--active"
          className="Header__nav-link"
          key="c"
          to="/c"
        >
          C
        </NavLink>
        <div
          className="Header__nav-spacer"
        />
        <ExternalLink
          className="Header__nav-link"
          to="https://blog.erbridge.co.uk"
        >
          Blog
        </ExternalLink>
      </div>
      <div
        className="Header__nav-menu-trigger"
        onClick={[Function]}
      >
        Menu
        <div
          className="Header__nav-menu-icon"
        />
      </div>
    </div>
  </div>
  <div
    className="Header__nav-menu"
  >
    <div
      className="Header__nav-menu__content"
    >
      <NavLink
        activeClassName="Header__nav-menu-link--active"
        className="Header__nav-menu-link"
        key="a"
        onClick={[Function]}
        tabIndex={-1}
        to="/a"
      >
        A
      </NavLink>
      <NavLink
        activeClassName="Header__nav-menu-link--active"
        className="Header__nav-menu-link"
        key="b"
        onClick={[Function]}
        tabIndex={-1}
        to="/b"
      >
        B
      </NavLink>
      <NavLink
        activeClassName="Header__nav-menu-link--active"
        className="Header__nav-menu-link"
        key="c"
        onClick={[Function]}
        tabIndex={-1}
        to="/c"
      >
        C
      </NavLink>
    </div>
  </div>
</Fragment>
`;

exports[`renders correctly with empty routes 1`] = `
<Fragment>
  <div
    className="Header"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="Header__content"
    >
      <NavLink
        exact={true}
        to="/"
      >
        <Logo
          className="Header__logo"
          focused={false}
        />
      </NavLink>
      <div
        className="Header__spacer"
      />
      <div
        className="Header__nav"
      >
        <ExternalLink
          className="Header__nav-link"
          to="https://blog.erbridge.co.uk"
        >
          Blog
        </ExternalLink>
      </div>
      <div
        className="Header__nav-menu-trigger"
        onClick={[Function]}
      >
        Menu
        <div
          className="Header__nav-menu-icon"
        />
      </div>
    </div>
  </div>
  <div
    className="Header__nav-menu"
  >
    <div
      className="Header__nav-menu__content"
    />
  </div>
</Fragment>
`;

exports[`renders correctly without optional props 1`] = `
<Fragment>
  <div
    className="Header"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="Header__content"
    >
      <NavLink
        exact={true}
        to="/"
      >
        <Logo
          className="Header__logo"
          focused={false}
        />
      </NavLink>
      <div
        className="Header__spacer"
      />
      <div
        className="Header__nav"
      >
        <ExternalLink
          className="Header__nav-link"
          to="https://blog.erbridge.co.uk"
        >
          Blog
        </ExternalLink>
      </div>
      <div
        className="Header__nav-menu-trigger"
        onClick={[Function]}
      >
        Menu
        <div
          className="Header__nav-menu-icon"
        />
      </div>
    </div>
  </div>
  <div
    className="Header__nav-menu"
  >
    <div
      className="Header__nav-menu__content"
    />
  </div>
</Fragment>
`;
